pkg/core/klinebuilder.go

```
@@ -0,0 +1,190 @@
+	numTrades                                          uint64
+}
+
+// NewKLineBuilder: constructor of kLineBuilder
+//   - symbol: symbol to trace on
+//   - minInterval: unit interval, related to your signal timeframe.
+//     All the supported intervals of the binding stream should be multiple of this interval.
+func NewKLineBuilder(symbol string, minInterval types.Interval) *kLineBuilder {
+	logger := logrus.WithField("symbol", symbol)
+	return &kLineBuilder{
+		symbol:      symbol,
+		minInterval: minInterval,
+		intervals:   make(map[types.Interval]struct{}),
+		klinesMap:   make(map[types.Interval]*types.KLine),
+		logger:      logger,
+	}
+}
+
+func (c *kLineBuilder) Subscribe(interval types.Interval) {
+	c.intervals[interval] = struct{}{}
+}
+
+func (c *kLineBuilder) BindStream(stream types.Stream) {
+	stream.OnMarketTrade(c.handleMarketTrade)
+}
+
+func (c *kLineBuilder) handleMarketTrade(trade types.Trade) {
+	// the market data stream may subscribe to trades of multiple symbols
+	// so we need to check if the trade is for the symbol we are interested in
+	if trade.Symbol != c.symbol {
+		return
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.price = trade.Price
+	c.close = trade.Price
+	if !c.high.IsZero() {
+		if c.price.Compare(c.high) > 0 {
+			c.high = c.price
+		}
+	} else {
+		c.high = c.price
+	}
+	if !c.low.IsZero() {
+		if c.price.Compare(c.low) < 0 {
+			c.low = c.price
+		}
+	} else {
+		c.low = c.price
+	}
+	if c.open.IsZero() {
+		c.open = c.price
+	}
+	c.volume = c.volume.Add(trade.Quantity)
+	c.quoteVolume = c.quoteVolume.Add(trade.QuoteQuantity)
+	c.numTrades++
+	c.updateTime = trade.Time
+}
+
+func (c *kLineBuilder) Run(ctx context.Context) {
+	if len(c.intervals) == 0 {
+		return
+	}
+	minDuration := c.minInterval.Duration()
+	for interval := range c.intervals {
+		if interval.Duration()%minDuration != 0 {
+			c.logger.Errorf("interval %s is not a multiple of minInterval %s", interval, c.minInterval)
+			panic(interval)
```

use fmt.Errorf and pass the error to panic instead of Errorf logging here

it's a logical exception (user configuration faults)

---

ok